api/src/main/java/io/minio/MinioClient.java

```diff
@@ -976,10 +978,15 @@ private Request createRequest(Method method, String bucketName, String objectNam
           String[] hashes = Digest.sha256Md5Hashes(data, len);
           sha256Hash = hashes[0];
           md5Hash = hashes[1];
-        } else if (method == Method.PUT && queryParamMap != null && queryParamMap.containsKey("lifecycle")) {
+        } else if (method == Method.PUT && queryParamMap != null &&  queryParamMap.containsKey("lifecycle"))  {
           String[] hashes = Digest.sha256Md5Hashes(data, len);
           sha256Hash = hashes[0];
           md5Hash = hashes[1];
+        } else if (method == Method.PUT && headerMap.containsKey("retention")) {
+          String[] hashes = Digest.sha256Md5Hashes(data, len);
+          sha256Hash = hashes[0];
+          md5Hash = hashes[1];
+          headerMap.remove("retention","");
```

Comment:
Not a good approach to pass and remove headers or params for sake of checksum calculation. You could pass a flag from higher level calls whether MD5 checksum calculation is required or not. You could add this improvement and remove these nested `if` guesses.